File: dynamic_listings/python/default/design_patterns/proxy.py
class User:
    # This class represents a user, with their permissions
    # ...
    def has_permission(self):
        # Changes according to user permissions
        # ...


class Request:
    # This class represents a web request
    def __init__(self, user: User):
        self.user = user


class WebPage:
    # Represents a call to a web page
    # This function gets implemented in the concrete classes;
    def get(self, request: Request):
        raise NotImplementedError


class WebPageProxy(WebPage):
    # Represents an authentication proxy for a web page
    def get(self, request: Request):
        # Get the requesting user
        requesting_user = request.user
        if (requesting_user.has_permission()):
            return super().get(request)
        else:
            return HttpResponse("This user cannot access this page")


if __name__ == '__main__':
    user = User()
    # The following line is valid due to polimorphism
    page = WebPageProxy()
    # If the user has permission the page will come through.
    page.get(Request(user))

File: dynamic_listings/rust/default/design_patterns/proxy.rs
# #![allow(dead_code)]
struct HttpResponse(&'static str);

struct User {
    // This class represents a user, with their permissions
    // ...
}

impl User {
#   fn new() -> Self { todo!() }
    pub fn can_do_thing(&self) -> bool {
        // Changes according to user permissions
        // ...
#       todo!()
    }
}

struct Request {
    // This class represents a web request
    user: User,
}

impl Request {
    pub fn new(user: User) -> Self {
        Self { user }
    }
}

trait WebPage {
    /*
     * Represents a call to a web page
     */
    // This function gets implemented in the concrete classes;
    fn get(&self, request: &Request) -> HttpResponse;
}

struct WebPageProxy(Box<dyn WebPage>);
# impl WebPageProxy {
#   fn new() -> Self { todo!() }
# }

impl WebPage for WebPageProxy {
    // Represents an authentication proxy for a web page
    fn get(&self, request: &Request) -> HttpResponse {
        // Get the requesting user
        let requesting_user = &request.user;
        if requesting_user.can_do_thing() {
            self.0.get(request)
        } else {
            HttpResponse("This user cannot access this page")
        }
    }
}

fn main() {
    let user = User::new();
    let page = WebPageProxy::new();
    // If the user has permission the page will come through.
    WebPage::get(&page, &Request::new(user));
}
